File: LeetCode/1306. Jump Game III/solution.cpp
/**
 * 56 / 56 test cases passed.
 * Runtime: 48 ms
 * Memory Usage: 30.6 MB 
 */
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        if (arr[start] == 0) return true;
        int n = arr.size();
        queue<int> que;
        que.emplace(start);
        vector<bool> used(n);
        while (!que.empty()) {
            auto curr = que.front(); que.pop();
            int next = curr + arr[curr];
            if (next < n && !used[next]) {
                if (arr[next] == 0) return true;
                used[next] = true;
                que.emplace(next); 
            }
            next = curr - arr[curr];
            if (next >= 0 && !used[next]) {
                if (arr[next] == 0) return true;
                used[next] = true;
                que.emplace(next); 
            }
        }
        return false;
    }
};

File: LeetCode/1306. Jump Game III/solution.py
"""
56 / 56 test cases passed.
Runtime: 44 ms
Memory Usage: 19.5 MB
"""
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        if arr[start] == 0:
            return True
        n = len(arr)
        que = collections.deque([start])
        vis = [False] * n
        while que:
            curr = que.pop()
            nxt = curr + arr[curr]
            if nxt < n and not vis[nxt]:
                if arr[nxt] == 0:
                    return True
                vis[nxt] = True
                que.append(nxt)
            nxt = curr - arr[curr]
            if nxt >= 0 and not vis[nxt]:
                if arr[nxt] == 0:
                    return True
                vis[nxt] = True
                que.append(nxt)
        return False


"""
56 / 56 test cases passed.
Runtime: 44 ms
Memory Usage: 19.5 MB
"""
class Solution2:
    def canReach(self, arr: List[int], start: int) -> bool:
        if 0 <= start < len(arr):
            if arr[start] == 0: return True
            temp = arr[start]
            arr[start] = len(arr)
            return self.canReach(arr, start + temp) or self.canReach(arr, start - temp)
        return False
